### DISCUSSION TITLE
Treesitter opts.ensure_installed

### DISCUSSION THREAD

**Discussion by azdanov** [LazyVim/LazyVim#1000] [Q&A]
Hey. I'm slowly learning more about neovim, lazy and lua 🙂.

Would someone be so kind as to help me out.

What would be the proper way to add new languages to treesitter opts?
I see 2 ways. What is the idea when comparing to a `"table"`?

`go.lua`
```lua
opts = function(_, opts)
  vim.list_extend(opts.ensure_installed, {
    "go",
    "gomod",
    "gowork",
    "gosum",
  })
end
```

`json.lua`
```lua
opts = function(_, opts)
  if type(opts.ensure_installed) == "table" then
    vim.list_extend(opts.ensure_installed, { "json", "json5", "jsonc" })
  end
end
```

Also a bonus question. When does this deduplication (vim.tbl_filter lambda) happen inside `treesitter.lua`? In the end when all other "extras" are loaded?
```lua
if type(opts.ensure_installed) == "table" then
  ---@type table<string, boolean>
  local added = {}
  opts.ensure_installed = vim.tbl_filter(function(lang)
    if added[lang] then
      return false
    end
    added[lang] = true
    return true
  end, opts.ensure_installed)
end
```

**Comment by amaanq** [comment]
nvim-ts allows "all" as an install option - which is not a table, hence the check.

the last one seems to just filter unique parsers
